### ISSUE TITLE
Servant binary weird behaviour when running in docker container

### ISSUE THREAD

**Issue by naglalakk** [haskell-servant/servant#1087]
I'm getting a really weird issue when running servant inside a docker container. After doing stack builld and stack install from my Dockerfile I run a supervisor process which is suppose to run the servant server. It runs but it's not accessible. What is super weird is that when I do "docker exec" into the container and start the server by running exactly the same command it works. 
It seems that the server is not accessible but running when docker runs this command. 

**How is this related to Servant?**
maybe it isn't, but this binary running the servant server is the only binary behaving this way.
I've also been running other haskell binaries with the same script (I have a worker running with hworker and this runs fine)

I'm testing this with the following supervisord config

```
[supervisord]
nodaemon=true

[program:server]
directory=/code
command=stack exec example-service
autostart=true
autorestart=true
stdout_logfile=/code/logs/supervisor-server.log

[program:testserver]
command=python -m SimpleHTTPServer 8082
autostart=true
autorestart=true
```

I run a python server on port 8082 and this is accessible. The example-service is suppose to be running on port 8081. It's not accessible until I manually start it within the container.

Here is the docker-compose command

```
        command:
            - ./config/wait-for-it.sh
            - db:5432
            - --
            - ./config/wait-for-it.sh
            - elasticsearch:9200
            - --
            - supervisord
            - -c
            - /code/config/supervisord.conf
```

** How is it not accessible? **
The example-server binary seems to be running when started with the docker-compose command above. However it's running but port 8081 is still not accessible. When I do ps aux I can see the example-server process is running.

```
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.1  51176 15316 ?        Ss   13:13   0:01 /usr/bin/python /usr/bin/supervisord -c /code/config/supervisord.conf
root        36  0.0  0.1  37932 12620 ?        S    13:13   0:01 python -m SimpleHTTPServer 8082
root        37  0.0  0.1 1073850116 16328 ?    S    13:13   0:00 /code/.stack-work/install/x86_64-linux/lts-10.5/8.2.2/bin/example-service
root       154  2.0  0.0  21948  3696 pts/0    Ss   15:09   0:00 /bin/bash
root       161  0.0  0.0  19188  2412 pts/0    R+   15:09   0:00 ps aux
```

I don't want to fill your issue tracker with unrelated bugs I just wanted to check if anyone had experienced the same and what might be the cause. I've been debugging this for hours now and I cannot see what might possibly be causing this.

**Comment by erewok** [comment]
I think it's highly unlike that this has anything to do with Servant. 

Looking at what you have here, it's unusual, in my experience, to run a web service inside a docker container with a process-manager such as supervisor. Typically, you run your process directly inside the container and load-balance traffic to your container from the outside with a load-balancer that also checks a `/health` endpoint, which will trigger container restarts if there are any issues (this is all assuming that you've exposed the proper port(s)).

In short, there are some atypical things about how you're running Servant inside a docker container, but there's no reason why Servant would have any particular issues running inside a docker container.

**Comment by naglalakk** [comment]
Hi both and thank you very much for your answers. @alpmestan confirmed I can run your snippet with supervisord and it works when docker starts it, I don't need to start it manually. So this indicates that it's probably something in my other Main file which handles running the servant server. 
I agree don't think this has anything to do with Servant. 
@erewok you are correct and usually I don't actually run it like this. But it was a part of the trial and error steps. The result is the same when running only the server command.
I'll close this issue so it's not making noise but keep you posted if I figure out this very weird bug.

**Comment by naglalakk** [comment]
Ok confirmed that this is solely my issue and has nothing to do with Servant at all. I'm picking up the port through an env variable and for some reason when running it with docker the port got assigned to 9200 instead of 8081. A bit weird because when I docker exec into it the PORT is set to 8081. But I just changed the name of the env variable and now all is good.
Thank you as always for outstanding work on this.
